### ISSUE TITLE
关于torch版本转为paddle版本，模型性能极差。

### ISSUE THREAD

**Issue by XiangLiu0731** [PaddlePaddle/Paddle#43196]
### bug描述 Describe the Bug

在将GuideNet网络pytorch版本转为paddle版本(2.3)后，模型基本不收敛。经检查，两个版本模型参数结构完全一致（除了BatchNorm2D）
在模型参数显示中，paddle参数为pytorch两倍，请问一下是显示问题还是我用法有问题。
下面是测试BatchNorm2D参数代码：
“”“
import paddle.nn as nn
model = paddle.Model(nn.BatchNorm2D(32))
model.summary((2, 32, 128, 128))
”“”
结果：

---------------------------------------------------------------------------
 Layer (type)       Input Shape          Output Shape         Param #    
===========================================================================
 BatchNorm2D-1  [[2, 32, 128, 128]]   [2, 32, 128, 128]         128      
===========================================================================

显示BatchNorm2D层参数为128，但是weight+bias不是只有64个参数吗，这样使用BatchNorm2D是正确的吗？

**Comment by XiangLiu0731** [comment]
我还注意到测试代码显示128个参数都是非训练的。
Total params: 128
Trainable params: 0
Non-trainable params: 128
请问如果想要像pytorch一样默认均值方差在训练时根据全部数据计算得到，而weight，bias在训练时更新应该怎么实现呢？

**Comment by wjj19950828** [comment]
@XiangLiu0731 您好，请问下您是如何从pytorch转到paddle呢，可以提供一下复现脚本吗~

**Comment by XiangLiu0731** [comment]
> 

我并没有使用脚本转源码，而是自己手动转的。我又统一了一下两个版本模型的参数初始化方式，经测试，相同输入条件下两个版本的前向传播结果是一致的，但是测试后向传播时，发现两个模型的梯度计算差异较大（约三个数量级）。测试代码如下：
pytorch版本梯度测试代码：
seed_everything(4321)
model=GuideNet().cuda()
model.train()
x = torch.from_numpy(np.random.randn(1, 1, 128, 128).astype('float32')).cuda()
y = torch.from_numpy(np.random.randn(1, 1, 128, 128).astype('float32')).cuda()
 z = torch.from_numpy(np.random.randn(1, 1, 128, 128).astype('float32')).cuda()
x.requires_grad = True
y.requires_grad = True
z.requires_grad = True
s1, s2 = model(x, y, z)
loss = nn.functional.mse_loss(s1, torch.ones_like(x))
loss.backward()
print(x.grad)

而paddle模型梯度测试代码
paddle.set_device("gpu")
paddle.disable_static()
seed_everything(4321)
model=GuideNet()
model.train()
x = paddle.to_tensor(np.random.randn(1, 1, 128, 128), dtype='float32')
y = paddle.to_tensor(np.random.randn(1, 1, 128, 128), dtype='float32')
z = paddle.to_tensor(np.random.randn(1, 1, 128, 128), dtype='float32')
x.stop_gradient = False
y.stop_gradient = False
z.stop_gradient = False
s1, s2 = model(x, y, z)
loss = paddle.nn.functional.mse_loss(s1, paddle.ones_like(x))
aaa = paddle.grad(outputs=[loss], inputs=[x, y])[0]  #s1只根据x和y计算获得
print(aaa)

**Comment by XiangLiu0731** [comment]
> @XiangLiu0731 您好，请问下您是如何从pytorch转到paddle呢，可以提供一下复现脚本吗~

如果您需要模型代码，您可以留一下邮箱，我将paddle和pytorch版本模型代码发您。

**Comment by XiangLiu0731** [comment]
经过对原模型各层梯度进行比较，我们最终定位到梯度出现在自定义算子梯度后一层（包含自定义算子模块处梯度与pytorch一致，但该模块结果输入resnet模块后，梯度发生错误）。paddle模型代码如下所示（标注出了模型梯度出现错误的地方，为了代码易读性，忽略了部分函数，如需原始代码，请留邮箱）。其中paddle版本为2.3.0， cuda=11.1。
此外，我们还尝试将包含自定义算子模块去除，此时整体网络梯度与pytorch一致。单独对包含自定义算子模块梯度进行测试，此时梯度仍与pytorch一致。个人怀疑自定义算子（C++ add CUDA）与后续网络梯度回传处出现问题。

````
class DeepGuideNet(nn.Layer):
    def __init__(self, args, block=BasicBlock, bc=16, img_layers=[2, 2, 2, 2, 2], depth_layers=[2, 2, 2, 2, 2],
                 norm_layer=nn.BatchNorm2D, guide=Guide, weight_ks=1):
        super().__init__()
        if args.freeze_fgnet:
            weight_attr = paddle.framework.ParamAttr(learning_rate=0.0, trainable=False)
        else:
            weight_attr = None

        self._FGNet = FGNet(weightattr=weight_attr)
        self._norm_layer = norm_layer

        self.conv_img = Basic2d(1, bc * 2, norm_layer=norm_layer, kernel_size=5, padding=2)
        in_channels = bc * 2
        self.inplanes = in_channels
        self.layer1_img = self._make_layer(block, in_channels * 2, img_layers[0], stride=2, norm=True)

        self.guide1 = guide(in_channels * 2, in_channels * 2, norm_layer, weight_ks)
        self.inplanes = in_channels * 2 * block.expansion
        self.layer2_img = self._make_layer(block, in_channels * 4, img_layers[1], stride=2, norm=True)

        self.guide2 = guide(in_channels * 4, in_channels * 4, norm_layer, weight_ks)
        self.inplanes = in_channels * 4 * block.expansion
        self.layer3_img = self._make_layer(block, in_channels * 8, img_layers[2], stride=2, norm=True)

        self.guide3 = guide(in_channels * 8, in_channels * 8, norm_layer, weight_ks)
        self.inplanes = in_channels * 8 * block.expansion
        self.layer4_img = self._make_layer(block, in_channels * 8, img_layers[3], stride=2, norm=True)

        self.guide4 = guide(in_channels * 8, in_channels * 8, norm_layer, weight_ks)
        self.inplanes = in_channels * 8 * block.expansion
        self.layer5_img = self._make_layer(block, in_channels * 8, img_layers[4], stride=2, norm=True)

        self.layer2d_img = Basic2dTrans(in_channels * 4, in_channels * 2, norm_layer)
        self.layer3d_img = Basic2dTrans(in_channels * 8, in_channels * 4, norm_layer)
        self.layer4d_img = Basic2dTrans(in_channels * 8, in_channels * 8, norm_layer)
        self.layer5d_img = Basic2dTrans(in_channels * 8, in_channels * 8, norm_layer)

        self.conv_lidar = Basic2d(1, bc * 2, norm_layer=None, kernel_size=5, padding=2)  # conv + relu
        self.inplanes = in_channels  # 32
        self.layer1_lidar = self._make_layer(block, in_channels * 2, depth_layers[0], stride=2, norm=True)  # 64, 1/2
        self.inplanes = in_channels * 2 * block.expansion
        self.layer2_lidar = self._make_layer(block, in_channels * 4, depth_layers[1], stride=2, norm=True)  # 128, 1/4
        self.inplanes = in_channels * 4 * block.expansion
        self.layer3_lidar = self._make_layer(block, in_channels * 8, depth_layers[2], stride=2, norm=True)  # 256, 1/8
        self.inplanes = in_channels * 8 * block.expansion
        self.layer4_lidar = self._make_layer(block, in_channels * 8, depth_layers[3], stride=2, norm=True)
        self.inplanes = in_channels * 8 * block.expansion
        self.layer5_lidar = self._make_layer(block, in_channels * 8, depth_layers[4], stride=2, norm=True)

        self.layer1d = Basic2dTrans(in_channels * 2, in_channels, norm_layer)
        self.layer2d = Basic2dTrans(in_channels * 4, in_channels * 2, norm_layer)
        self.layer3d = Basic2dTrans(in_channels * 8, in_channels * 4, norm_layer)
        self.layer4d = Basic2dTrans(in_channels * 8, in_channels * 8, norm_layer)
        self.layer5d = Basic2dTrans(in_channels * 8, in_channels * 8, norm_layer)

        self.conv_for1 = nn.Conv2D(bc * 2, 1, kernel_size=1, stride=1, padding=0)
        self.ref_for1 = block(bc * 2, bc * 2, norm_layer=norm_layer, act=False)

        self._initialize_weights()  # 初始化权重，保证初始化模型参数与pytorch版本一致

    def forward(self, gray_img, denselidar, sparselidar):
        fg_pred = self._FGNet(gray_img, sparselidar)

        c0_img = self.conv_img(gray_img)  # conv+bn+relu -->[B, 32, H, W]
        c1_img = self.layer1_img(c0_img) 
        c2_img = self.layer2_img(c1_img) 
        c3_img = self.layer3_img(c2_img) 
        c4_img = self.layer4_img(c3_img)  
        c5_img = self.layer5_img(c4_img)  # [B, 256, H/32, W/32]
        dc5_img = self.layer5d_img(c5_img)  # [B, ]
        c4_mix = dc5_img + c4_img  # [B, 128, H/16, W/16]
        dc4_img = self.layer4d_img(c4_mix)
        c3_mix = dc4_img + c3_img
        dc3_img = self.layer3d_img(c3_mix)
        c2_mix = dc3_img + c2_img
        dc2_img = self.layer2d_img(c2_mix)
        c1_mix = dc2_img + c1_img

        c0_lidar = self.conv_lidar(denselidar)
        c1_lidar = self.layer1_lidar(c0_lidar)
        c1_lidar_dyn = self.guide1(c1_lidar, c1_mix)  # 该层包含了自定义算子，经测试，该层算子梯度与pytorch一致

        # ----------截至目前为止，c1_lidar_dyn 梯度未出现错误，与pytorch版本一致，证明自定义算子梯度计算正确---------
        c2_lidar = self.layer2_lidar(c1_lidar_dyn)
        # -------------------------------------------------c2_lidar 梯度开始出现错误-------------------------------------------------
        
        c2_lidar_dyn = self.guide2(c2_lidar, c2_mix)
        c3_lidar = self.layer3_lidar(c2_lidar_dyn)
        c3_lidar_dyn = self.guide3(c3_lidar, c3_mix)
        c4_lidar = self.layer4_lidar(c3_lidar_dyn)
        c4_lidar_dyn = self.guide4(c4_lidar, c4_mix)
        c5_lidar = self.layer5_lidar(c4_lidar_dyn)
        c5 = c5_img + c5_lidar

        dc5 = self.layer5d(c5)  # 256, 1/16
        c4 = dc5 + c4_lidar_dyn
        dc4 = self.layer4d(c4)  # 256, 1/8
        c3 = dc4 + c3_lidar_dyn
        dc3 = self.layer3d(c3)  # 128, 1/4
        c2 = dc3 + c2_lidar_dyn
        dc2 = self.layer2d(c2)  # 64, 1/2
        c1 = dc2 + c1_lidar_dyn
        dc1 = self.layer1d(c1)  # 32, 1/1
        c0 = dc1 + c0_lidar

        feats_1 = self.ref_for1(c0)
        out_1 = self.conv_for1(feats_1)
        return out_1, fg_pred
```

**其中较为重要的自定义算子层定义如下**：

class Basic2dLocal(nn.Layer):
    def __init__(self, out_channels, norm_layer=None):
        super().__init__()
        if norm_layer is None:
            norm_layer = nn.BatchNorm2D

        self.convlocal = Conv2dLocal  # 自定义算子操作，输入为[B, C, H, W]与[B, C*9, H, W]，输出为[B, C, H, W]

        self.bn = norm_layer(out_channels)
        self.relu = nn.ReLU()

    def forward(self, input, weight):
        out = self.convlocal(input, weight)
        out = self.bn(out)
        out = self.relu(out)
        return out

**总体网络中Guide层定义如下**：

class Guide(nn.Layer):
    def __init__(self, input_planes, weight_planes, norm_layer=None, weight_ks=3):
        super().__init__()
        if norm_layer is None:
            norm_layer = nn.BatchNorm2D
        self.local = Basic2dLocal(input_planes, norm_layer)
        self.pool = nn.AdaptiveAvgPool2D((1, 1))
        self.conv11 = Basic2d(input_planes + weight_planes, input_planes, None)

        self.conv12 = nn.Conv2D(input_planes, input_planes * 9, kernel_size=weight_ks, padding=weight_ks // 2)
        self.conv21 = Basic2d(input_planes + weight_planes, input_planes, None)

        self.conv22 = nn.Conv2D(input_planes, input_planes * input_planes, kernel_size=1, padding=0)
        self.br = nn.Sequential(
            norm_layer(num_features=input_planes),
            nn.ReLU(),
        )
        self.conv3 = Basic2d(input_planes, input_planes, norm_layer)

    def forward(self, input, weight):
        B, Ci, H, W = input.shape
        weight = paddle.concat([input, weight], axis=1)  # [B, 2C, H, W]
        weight11 = self.conv11(weight)
        weight12 = self.conv12(weight11)  # [B, 9c, H, W]

        weight21 = self.conv21(weight)
        weight21 = self.pool(weight21)
        weight22 = paddle.reshape(self.conv22(weight21), [B, -1, Ci])  # [B, C, C]

        out = paddle.reshape(self.local(input, weight12), [B, Ci, -1])

        out = paddle.reshape(paddle.bmm(weight22, out), [B, Ci, H, W])
        out = self.br(out)
        out = self.conv3(out)
        return out

**Comment by XiangLiu0731** [comment]
> @XiangLiu0731 您好，请问下您是如何从pytorch转到paddle呢，可以提供一下复现脚本吗~

目前我们使用了paddle自带api替代了自定义算子操作，并且获得了与pytorch一致的梯度，但仍希望您能够解决上述发现的问题。此外，希望paddle框架在windows平台下能够早日支持多线程读取数据功能！
